src/Common/Allocator.cpp

```diff
@@ -144,8 +181,54 @@ void * Allocator<clear_memory_, populate>::realloc(void * buf, size_t old_size,
     {
         /// nothing to do.
         /// BTW, it's not possible to change alignment while doing realloc.
+        return buf;
     }
-    else if (alignment <= MALLOC_MIN_ALIGNMENT)
+
+#if USE_GWP_ASAN
+    if (unlikely(GWPAsan::GuardedAlloc.shouldSample()))
+    {
+        if (void * ptr = GWPAsan::GuardedAlloc.allocate(new_size, alignment))
+        {
+            auto trace_free = CurrentMemoryTracker::free(old_size);
+            auto trace_alloc = CurrentMemoryTracker::alloc(new_size);
+            trace_free.onFree(buf, old_size);
+
+            memcpy(ptr, buf, std::min(old_size, new_size));
+            free(buf, old_size);
+            trace_alloc.onAlloc(buf, new_size);
+
+            if constexpr (clear_memory)
+                if (new_size > old_size)
+                    memset(reinterpret_cast<char *>(ptr) + old_size, 0, new_size - old_size);
+
+            if constexpr (populate)
+                prefaultPages(ptr, new_size);
+
+            ProfileEvents::increment(ProfileEvents::GWPAsanAllocateSuccess);
+            return ptr;
+        }
+        else
+        {
+            ProfileEvents::increment(ProfileEvents::GWPAsanAllocateFailed);
+        }
+    }
+
+    if (unlikely(GWPAsan::GuardedAlloc.pointerIsMine(buf)))
+    {
+        /// Big allocs that requires a copy. MemoryTracker is called inside 'alloc', 'free' methods.
+        void * new_buf = alloc(new_size, alignment);
+        memcpy(new_buf, buf, std::min(old_size, new_size));
+        free(buf, old_size);
```

Comment:
buf which belongs to gwp asan should be freed by ` GWPAsan::GuardedAlloc.deallocate(buf);`